service/matching/matching_engine_test.go

```diff
@@ -2499,6 +2497,100 @@ func (s *matchingEngineSuite) TestUnknownBuildId_Match() {
 	wg.Wait()
 }
 
+func (s *matchingEngineSuite) TestUnknownBuildId_Demoted_Match() {
+	namespaceId := namespace.ID(uuid.New())
+	tq := "makeToast"
+	unknown := "unknown"
+	build1 := "build1"
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	s.mockMatchingClient.EXPECT().UpdateWorkerBuildIdCompatibility(gomock.Any(), &matchingservice.UpdateWorkerBuildIdCompatibilityRequest{
+		NamespaceId: namespaceId.String(),
+		TaskQueue:   tq,
+		Operation: &matchingservice.UpdateWorkerBuildIdCompatibilityRequest_PersistUnknownBuildId{
+			PersistUnknownBuildId: unknown,
+		},
+	}).Return(&matchingservice.UpdateWorkerBuildIdCompatibilityResponse{}, nil).AnyTimes() // might get called again on dispatch from spooled
+
+	// add a task for an unknown build id, will get redirected to guessed set
+	_, err := s.matchingEngine.AddWorkflowTask(ctx, &matchingservice.AddWorkflowTaskRequest{
+		NamespaceId:      namespaceId.String(),
+		Execution:        &commonpb.WorkflowExecution{RunId: "run", WorkflowId: "wf"},
+		ScheduledEventId: 123,
+		TaskQueue:        &taskqueuepb.TaskQueue{Name: tq, Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+		VersionDirective: &taskqueue.TaskVersionDirective{
+			Value: &taskqueue.TaskVersionDirective_BuildId{
+				BuildId: unknown,
+			},
+		},
+	})
+	s.NoError(err)
+	// allow taskReader to finish starting dispatch loop so we don't get an extra load
```

Comment:
Can you clarify what you mean here? Also, is it possible to avoid the sleep here?

Reply:
edited the comment.

if it was possible to avoid the sleep I wouldn't have added it :)  it passes > 95% without it, but this gets it > 99.9%

it's a similar stop-before-start-finishes problem, where if I try to stop it before taskReader has loaded + tried to dispatch a task, it'll do that and mess up the unload that I'm trying to do.